misc/config_tools/launch_config/launch_cfg_gen.py

```diff
@@ -262,9 +262,14 @@ def generate_for_one_vm(board_etree, hv_scenario_etree, vm_scenario_etree, vm_id
     if eval_xpath(vm_scenario_etree, ".//console_vuart/text()") == "PCI":
         script.add_virtual_device("uart", options="vuart_idx:0")
 
-    for idx, conn in enumerate(eval_xpath_all(hv_scenario_etree, f".//vuart_connection[endpoint/vm_name = '{vm_name}']"), start=1):
-        if eval_xpath(conn, "./type/text()") == "pci":
-            script.add_virtual_device("uart", options=f"vuart_idx:{idx}")
+    for idx, conn in enumerate(eval_xpath_all(hv_scenario_etree, f"//vuart_connection[endpoint/vm_name/text() = '{vm_name}']"), start=1):
+        if eval_xpath(conn, f"./type/text()") == "pci":
+            vbdf = eval_xpath(conn, f"./endpoint[vm_name/text() = '{vm_name}']/vbdf/text()")
+            if vbdf is not None:
+                slot = int((vbdf.split(":")[1].split(".")[0]), 16)
+            else:
+                slot = None
+            script.add_virtual_device("uart", slot, options=f"vuart_idx:{idx}")
```

Comment:
You need also let the BDF allocator be aware of the slots consumed by the vuart so that other devices are not using the same.

Reply:
> You need also let the BDF allocator be aware of the slots consumed by the vuart so that other devices are not using the same.

These code used to make sure the BDF is the same with user 's setting for post launched.

The BDF settings not just from allocator, also from user setting, so we need add aware logic and unified assert to check the scenario .

Reply:
I'm not referring to the static allocator, but the VirtualBDFAllocator class in launch_cfg_gen.py. An object of that class is used to assign vBDF to the virtual devices of a VM, and we need to make sure that it never attempts to use one a user has already reserved for vuart.

Reply:
I have update the VirtualBDFAllocator.